### PR TITLE
Add BiGRU

### PR DESCRIPTION
**Pull request by Eleven1Liu**
**Description**
1.  Add BiGRU and configs.
2. New metrics: RP@k and nDCG@k.

### CODE REVIEW

File: libmultilabel/metrics.py [hunk 1]
```diff
@@ -1,9 +1,72 @@
...
+        for y_pred, y_true in zip(preds, target):
+            self.score += self.ranking_rprecision_score(y_pred, y_true)
+        self.num_sample += len(preds)
+
+    def compute(self):
+        return self.score.float() / self.num_sample
+
+    def ranking_rprecision_score(self, y_pred, y_true):
+        n_pos = torch.sum(y_true == 1)
+        order = torch.argsort(-y_pred)
```
Comment (cjlin1) [inline libmultilabel/metrics.py:30]: torchmetrics has something called select_topk??
Also I don't think this function is needed. At line 22 we should avoid a loop and handle the batch together??

Comment (Eleven1Liu) [reply libmultilabel/metrics.py]: Yes. I followed lmtc's. Now there's no for loop in `RPrecision`.

File: example_config/EUR-Lex/bigru.yml [hunk 2]
```diff
@@ -0,0 +1,32 @@
+# data
+data_dir: data/EUR-Lex-v2
```
Comment (cjlin1) [inline example_config/EUR-Lex/bigru.yml:2]: for different data I think we should use another directory in example_config?

File: libmultilabel/model.py [hunk 3]
```diff
@@ -147,12 +147,21 @@ def test_epoch_end(self, step_outputs):
...
-        return self.eval_metric.update(batch_parts['pred_scores'], batch_parts['target'])
+        batch_size, num_classes = batch_parts['target'].shape
+        indexes = torch.arange(
+            batch_size*batch_parts['batch_idx'], batch_size*(batch_parts['batch_idx']+1))
+        indexes = indexes.unsqueeze(1).repeat(1, num_classes)
+        return self.eval_metric.update(
+            preds=batch_parts['pred_scores'],
+            target=batch_parts['target'],
+            indexes=indexes
+        )
```
Comment (cjlin1) [inline libmultilabel/model.py:164]: not clear about changes from 156-164. Maybe some discussion tomorrow

File: libmultilabel/networks/bigru.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+from ..networks.base import BaseModel
+
+
+class BiGRU(BaseModel):
+    def __init__(
+        self,
+        embed_vecs,
+        num_classes,
+        rnn_dim=512,
+        num_layers=1,
```
Comment (cjlin1) [inline libmultilabel/networks/bigru.py:17]: Now I feel that rnn_layers may be better. num_layers sounds like the number of layers of the whole network (including attention and linear etc.)

File: libmultilabel/networks/bigru.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+        x = pad_packed_sequence(x)[0]
+        x = x.permute(1,0,2)
+
+        x = torch.tanh(x)
+
+        # Apply per-label attention
+        alpha = torch.softmax(self.U.weight.matmul(x.transpose(1,2)), dim=2)
+
+        # Document representations are weighted sums using the attention
+        m = alpha.matmul(x) # (batch_size, label size, rnn_dim)
```
Comment (cjlin1) [inline libmultilabel/networks/bigru.py:55]: label size -> num_classes?

File: libmultilabel/metrics.py [hunk 6]
```diff
@@ -1,13 +1,43 @@
 import re
 
+import torch
 import numpy as np
-from torchmetrics import Metric, MetricCollection, F1, Precision, Recall
+from torchmetrics import Metric, MetricCollection, F1, Precision, Recall, RetrievalNormalizedDCG
+from torchmetrics.utilities.data import select_topk
+
+
+class RPrecision(Metric):
```
Comment (cjlin1) [inline libmultilabel/metrics.py:9]: RP@K isn't commonly used. Need some simple comments

File: libmultilabel/model.py (hunk 3 above)
Comment (Eleven1Liu) [reply libmultilabel/model.py:164]: Note: add comments

File: example_config/EUR-Lex-v2/bigru.yml [hunk 7]
```diff
@@ -0,0 +1,32 @@
+# data
+data_dir: data/EUR-Lex-v2
```
Comment (cjlin1) [inline example_config/EUR-Lex-v2/bigru.yml:2]: -v2 is our own name.. I think they call it Eur-Lex-57k. Also that's the name we used in libsvm data.

File: libmultilabel/networks/bigru.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+        m = alpha.matmul(x) # (batch_size, num_classes, rnn_dim)
+
+        # Compute a probability for each label
+        x = self.final.weight.mul(m).sum(dim=2).add(self.final.bias) # (batch_size, num_classes)
+
+        return {'logits': x[indices], 'attention': alpha[indices]}
+
+    def sort_data_by_length(self, text, length):
+        """Sort data by length. This function is the deterministic implementation of
+        `pack_padded_sequence` with `enforce_sorted = False`.
```
Comment (cjlin1) [inline libmultilabel/networks/bigru.py:72]: this comment is not good. It's not the implementation of "pack_padded..." because we don't pack sequence here (or say we still call pack_padd... at line 53). Should be something like

If data not sorted before calling ??, under enforce_sorted=False, ?? use_deterministic_algorithms must be False. To keep the use_deterministic_algorithms setting, here we sort ... before calling ??.

File: libmultilabel/networks/bigru.py [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+            text (torch.Tensor): Text sorted by lengths in descending order.
+            length (torch.Tensor): Lengths sorted in descending order.
+            indices (torch.Tensor): The indexes of the elements in the original data.
+        """
+        length = torch.as_tensor(length, dtype=torch.int64)
+        length, sorted_indices = torch.sort(length, descending=True)
+        sorted_indices = sorted_indices.to(text.device)
+        text = text.index_select(0, sorted_indices)
+
+        _, indices = torch.sort(sorted_indices) # get the original order
```
Comment (cjlin1) [inline libmultilabel/networks/bigru.py:88]: doing two sort isn't good. For the second sort, cannot we do some kind of index assignment. See the example below

>>> x = torch.randn(1,7)
>>> [z, ind] = torch.sort(x)
>>> y = torch.randn(1,7)
>>> ind
tensor([[3, 5, 6, 4, 1, 2, 0]])
>>> x
tensor([[ 1.7054,  0.3207,  1.5532, -2.0915,  0.2383, -0.4152,  0.0647]])
>>> y[0,3] = 0
>>> y[0,5] = 1
>>> y[0,6] = 2
>>> y[0,4] = 3
>>> y[0,1] = 4
>>> y[0,2] = 5
>>> y[0,0] = 6
>>> y
tensor([[6., 4., 5., 0., 3., 1., 2.]])
>>> 

though I don't know if we can do the above y assignment in just one line. I think it's possiblem

Comment (Eleven1Liu) [reply libmultilabel/networks/bigru.py:88]: Sure. This works!

```python
y = torch.clone(ind)
y.index_put_(tuple(ind.unsqueeze(0)), torch.arange(ind.size(-1)))
```

File: libmultilabel/model.py [hunk 10]
```diff
@@ -147,12 +144,23 @@ def test_epoch_end(self, step_outputs):
...
+        return {'batch_idx': batch_idx,
+                'loss': loss,
                 'pred_scores': torch.sigmoid(pred_logits),
                 'target': batch['label']}
 
     def _shared_eval_step_end(self, batch_parts):
-        return self.eval_metric.update(batch_parts['pred_scores'], batch_parts['target'])
+        batch_size, num_classes = batch_parts['target'].shape
+        # `indexes` is like sample ids that help `RetrievalNormalizedDCG` to calculate
+        # the NDCG score per sample.
```
Comment (cjlin1) [inline libmultilabel/model.py:155]: this comment isn't clear to me. Let's discuss more on Th

File: libmultilabel/networks/bigru.py [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+            indices (torch.Tensor): The indexes of the elements in the original data.
+        """
+        length = torch.as_tensor(length, dtype=torch.int64)
+        length, sorted_indices = torch.sort(length, descending=True)
+        sorted_indices = sorted_indices.to(data.device)
+        data = data.index_select(0, sorted_indices)
+
+        indices = torch.clone(sorted_indices)
+        indices.index_put_(tuple(sorted_indices.unsqueeze(0)),
+                           torch.where(sorted_indices >= 0)[0])
```
Comment (cjlin1) [inline libmultilabel/networks/bigru.py:92]: Does this where really work?

>>> torch.where(ind >= 0)[0]
tensor([0, 0, 0, 0, 0, 0, 0])
>>> torch.arange(ind.size(-1))
tensor([0, 1, 2, 3, 4, 5, 6])

do they have a more readable way to generate [0, 1, 2, 3, 4, 5, 6] ??

Comment (Eleven1Liu) [reply libmultilabel/networks/bigru.py:92]: Yes. Both (1) and (2) give`[0, 1, 2, 3, 4]`
**(1) torch.where**
```python
ind = torch.Tensor([ 1, 3, 4, 0, 2])
torch.where(ind >= 0)[0] # device='cuda:0'
``` 
**(2) torch.arange**
```python
torch.arange(ind.size(-1))) # device='cpu'
torch.arange(ind.size(-1))).to(ind) # device='cuda:0'
```

Comment (cjlin1) [reply libmultilabel/networks/bigru.py:92]: I see. Readability of  where(ind >= 0) is not good. Using arange is better. Is there a way to avoid the device issue?

File: libmultilabel/networks/bigru.py [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+            indices (torch.Tensor): The indexes of the elements in the original data.
+        """
+        length = torch.as_tensor(length, dtype=torch.int64)
+        length, sorted_indices = torch.sort(length, descending=True)
+        sorted_indices = sorted_indices.to(data.device)
+        data = data.index_select(0, sorted_indices)
+
+        indices = torch.clone(sorted_indices)
+        indices.index_put_(indices=tuple(sorted_indices.unsqueeze(0)),
+                           values=torch.arange(sorted_indices.size(-1)).to(sorted_indices))
```
Comment (Eleven1Liu) [inline libmultilabel/networks/bigru.py:92]: note: sorted_indices.device
dim=..(-1)

**Comment by Eleven1Liu** [comment]
> But I don't think this is right. Now sort_indices isn't used at all. I think we want
> indices[sorted_indices] = torch.arange...

This is the old one. We have this:
```python
data_size = sorted_indices.size(-1)
indices = torch.empty(data_size, dtype=torch.long)
indices[sorted_indices] = torch.arange(data_size)
```

at the latest commit.

**Review by cjlin1 — APPROVED** [review]
